tests/Unit/InternalClientTest.php

```diff
@@ -0,0 +1,292 @@
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack]);
+        $response = $client->sendRequest(new Request('GET', 'credits'));
+
+        self::assertSame([
+            'remaining' => '123',
+            'consumed' => '456',
+        ], $response);
+
+        self::assertCount(3, $requestHistory);
+        $request = $requestHistory[0]['request'];
+        self::assertInstanceOf(Request::class, $request);
+        self::assertEquals('https://example.com/credits', $request->getUri()->__toString());
+        $request = $requestHistory[1]['request'];
+        self::assertEquals('https://example.com/credits', $request->getUri()->__toString());
+        $request = $requestHistory[2]['request'];
+        self::assertEquals('https://example.com/credits', $request->getUri()->__toString());
+    }
+
+    public function testRetryFailure(): void
+    {
+        $responses = [];
+        for ($i = 0; $i < 30; $i++) {
+            $responses[] = new Response(
+                500,
+                ['Content-Type' => 'application/json'],
+                '{"message" => "Out of order"}'
+            );
+        }
+        $mock = new MockHandler($responses);
+        // Add the history middleware to the handler stack.
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack, 'backoffMaxTries' => 1]);
+        try {
+            $client->sendRequest(new Request('GET', 'credits'));
+            self::fail('Must throw exception');
+        } catch (BillingException $e) {
+            self::assertStringContainsString('500 Internal Server Error', $e->getMessage());
+        }
+        self::assertCount(2, $requestHistory);
+    }
+
+    public function testRetryFailureReducedBackoff(): void
+    {
+        $responses = [];
+        for ($i = 0; $i < 30; $i++) {
+            $responses[] = new Response(
+                500,
+                ['Content-Type' => 'application/json'],
+                '{"message" => "Out of order"}'
+            );
+        }
+        $mock = new MockHandler($responses);
+        // Add the history middleware to the handler stack.
+        $requestHistory = [];
+        $history = Middleware::history($requestHistory);
+        $stack = HandlerStack::create($mock);
+        $stack->push($history);
+        $client = $this->getClient(['handler' => $stack, 'backoffMaxTries' => 3]);
+        try {
+            $client->sendRequest(new Request('GET', 'credits'));
+            self::fail('Must throw exception');
+        } catch (BillingException $e) {
+            self::assertStringContainsString('500 Internal Server Error', $e->getMessage());
+        }
+        self::assertCount(4, $requestHistory);
```

Comment:
doufám, že je to copy & paste :)

Reply:
ano.. neco vypada ne-OK? je tam 3x retry = udela 4 requesty mi prijde cajk

Reply:
myslím celý ty testy

Reply:
jo, cely testy jsou c&p